### PR TITLE
For #11755 - Update coroutines to 1.6.1. Update tests.

### PR DESCRIPTION
**Pull request by Mugurell**
Started fresh since it wasn't easy to follow what still needed to be migrated over Grisha's patch from https://github.com/mozilla-mobile/android-components/pull/11516/files which also contained some extraneous changes - not needed for this upgrade.

Based the work mainly on https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/MIGRATION.md
Went with favoring `UnconfinedTestDispatcher` mainly because it eagerly enters `launch` and `async` blocks and so the migration process should be a bit less painful while our tests don't really on the exact order of execution to require the `StandardTestDispatcher` and incremental virtual time advancements.

------

To future reviewers: the patch though big is meant to be straightforward:
- goal was to refactor out `TestCoroutineScope`, `TestCoroutineDispatcher` and `runBlockingTest` since they are deprecated and would fail the build (if not suppressing the deprecation)
- as recommended I also tried to remove as many `runBlocking` usages as possible but in some cases with more intricate scenarios where interacting with other dispatchers / threads was needed some `runBlocking` calls were kept.
- `runTest` replaces all `runBlocking` and `runBlockingTest` usages.
`TestScope`, `UnconfinedTestDispatcher` and `StandardTestDispatcher` replaces all usages of `TestCoroutineScope`, `TestCoroutineDispatcher` as per the official documentation.
- `runTestOnMain` was used where we'd set the main dispatcher to be able to syncronize runs/code in tests/system under test. Occasionally when the `runTest` replacement would have the test fail I added the `MainCoroutineTest` rule and saw that the test passed with also using `runTestOnMain`.
- there are no modifications in the code being tested and no modifications in the tests logic. This seems like a good guarantee that if all tests pass on CI then the changes are good.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * - newly created coroutines inside it will automatically be reparented to the test coroutine context.
+ * - leveraging an already set strategy for entering launch / async blocks.
+ * - easier scheduling control.
+ *
+ * @see runTest
+ * @see Dispatchers.setMain
+ * @see TestDispatcher
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+fun runTestOnMain(
```
Comment (gabrielluong) [inline components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt:41]: This is a drive-by, but it wasn't easy to find the definition of this in the GitHub review since it forces you to load all the diffs because of the 195 file changes. Couldn't we have name this `runBlocking` to save ourselves a lot of refactoring?

Comment (Mugurell) [reply components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt:41]: My goal as per https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/MIGRATION.md was to refactor out `runBlocking` and `runBlockingTest`.
Seeing again these mentioned in tests would seem like a big code smell for me. 
And in that I think this would actually slow down development and review - just because we see something that should't be there. But maybe has reason to be there.

`runTestOnMain` is clearer in it's intent. It is intended to only be used when `Dispatchers.setMain` was previously called, otherwise it will throw an exception and fail the test.
It is just a quicker way of writing `mainCoroutineRule.scope.runTest` or `runTest(mainCoroutineRule.testDispatcher.scheduler)`.
In the majority of cases the main dispatcher is not set and `runTest` would work just fine.
Then there are the exceptional cases where `runBlocking` needs to be used - when `need to interact with dispatchers that execute on other threads, like Dispatchers.IO or Dispatchers.Default`.

**Comment by Mugurell** [comment]
Updated the patch to address some `AbstractFetchDownloadServiceTest` tests failing.
Was a bit tricky to find the cause and a fix. In the end it seemed to boil down to the intricate usage of various coroutines scopes and dispatchers and the fact that [AbstractFetchDownloadService continously tries to update the notifications from MainScope](https://github.com/mozilla-mobile/android-components/blob/c7eefe7d6c8913198b9656f304b2f91deb213519/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L301) - this stopping the tests using the same dispatcher from finishing.
So my approach on fixing this was to use two dispatchers, one in which to run the tests and one to set as main that would also be used to advance the virtual time.
Still had to keep some `runBlocking` usages here since they'll test how "real" time passed by checking agains `System.currentTimeMillis`.

**Comment by Mugurell** [comment]
Left a comment in the first post about my reasoning when approaching this and my expectations of the result.

**Comment by Mugurell** [comment]
`build-feature-downloads` failed because of:
```
java.lang.AssertionError: expected:<2> but was:<1>
at mozilla.components.feature.downloads.AbstractFetchDownloadServiceTest$cancelled download does not prevent other notifications$1.invokeSuspend(AbstractFetchDownloadServiceTest.kt:1735)
```
Ran these tests locally multiple times and got another to fail similarly.
Seems like if the test uses a different dispatcher to avoid waiting for main then it may finish too soon.
Will revert to using `runBlocking` as it was before for the download tests which rely on advancing the virtual time.

File: components/browser/icons/src/test/java/mozilla/components/browser/icons/generator/DefaultIconGeneratorTest.kt [hunk 2]
```diff
@@ -42,8 +44,9 @@ class DefaultIconGeneratorTest {
         assertNotEquals(0, generator.pickColor(res, ""))
     }
 
+    @ExperimentalCoroutinesApi
     @Test
-    fun generate() = runBlocking {
+    fun generate() = runTest(UnconfinedTestDispatcher()) {
```
Comment (csadilek) [inline components/browser/icons/src/test/java/mozilla/components/browser/icons/generator/DefaultIconGeneratorTest.kt:49]: Why do we need the `UnconfinedTestDispatcher` here relative to just using `runTest` as everywhere above.

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt [hunk 3]
```diff
@@ -434,7 +431,6 @@ class WebExtensionBrowserMenuTest {
         val menu: WebExtensionBrowserMenu = mock()
 
         menuItem.bind(menu, view)
-        testDispatcher.advanceUntilIdle()
```
Comment (csadilek) [inline components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt:437]: Where these unneeded? I see that below in `WebExtensionBrowserMenuItemTest` you kept the calls.

File: components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt [hunk 4]
```diff
@@ -1074,14 +1074,17 @@ class AbstractFetchDownloadServiceTest {
...
+        mainDispatcher.scheduler.runCurrent()
 
         // one of the notifications it is the group notification only for devices the support it
         assertEquals(2, shadowNotificationService.size())
     }
 
     @Test
     fun `notification is shown when download status is COMPLETED`() = runBlocking {
+        // val dispatcher = UnconfinedTestDispatcher()
+        // Dispatchers.setMain(dispatcher)
```
Comment (csadilek) [inline components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt:1087]: nit: leftover?

File: docs/changelog.md [hunk 5]
```diff
@@ -11,6 +11,10 @@ permalink: /changelog/
 * [Gecko](https://github.com/mozilla-mobile/android-components/blob/main/buildSrc/src/main/java/Gecko.kt)
 * [Configuration](https://github.com/mozilla-mobile/android-components/blob/main/.config.yml)
 
+* **support-test**
+  * ⚠️ **This is a breaking change**: `MainCoroutineRule` constructor now takes a `TestDispatcher` instead of deprecated `TestCoroutineDispatcher`. Default is `UnconfinedTestDispatcher`.
+  * ⚠️ **This is a breaking change**: `MainCoroutineRule.runBlockingTest` is replaced with a `runTestOnMain` top level function. . This method is preferred over the global `runTest` because it reparents new child coroutines to the test coroutine context.
```
Comment (csadilek) [inline docs/changelog.md:16]: Heads-up that this is used in Fenix: https://github.com/mozilla-mobile/fenix/blob/main/app/src/test/java/org/mozilla/fenix/settings/deletebrowsingdata/DeleteAndQuitTest.kt#L14 and will need a change.

**Review by csadilek — COMMENTED** [review]
Thanks for working through this upgrade!

I think the change to use `runTestOnMain` combined with our `MainCoroutineRule` makes sense and is straight forward.

I don't quite understand when to use the plain `runTest` (defaulting to an empty context) vs. `runTest(UnconfindedTestDispatcher()`? Added some comments below.

I think we should run this PR a few times before landing to make sure we're not introducing any intermittent test failures.

**Comment by csadilek** [comment]
The one test failure is unrelated to the changes and already filed: https://github.com/mozilla-mobile/android-components/issues/11342

File: components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt (hunk 4 above)
Comment (Mugurell) [reply components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt:1087]: Seems like, will remove this, thank you!

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt (hunk 3 above)
Comment (Mugurell) [reply components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt:437]: Nice catch!
Cannot say I remember this exactly but looking at it now the interesting part is the [`bind` method](https://github.com/mozilla-mobile/android-components/blob/6c0fe91ab83614265f533b819bf406fc046b2227/components/browser/menu/src/main/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItem.kt#L52) which gets to execute `setupIcon` - launching  a new coroutine to set an icon to the item:
```
MainScope().launch {
    loadIcon(view.context, imageView.measuredHeight)?.let {
        iconTintColorResource?.let { tint -> imageView.setTintResource(tint) }
        imageView.setImageDrawable(it)
    }
}
```

The two tests from `WebExtensionBrowserMenuTest` from where I removed `advanceUntilIdle` check only how many items are shown and what happens when they are clicked without the need to actually complete the `bind` method, or other similar setups since their result is not checked. 
The tests from `WebExtensionBrowserMenuItemTest` do granular checks of the results from executing `bind(..)` really only just two tests probably needing the `advanceUntilIdle` calls - where the tint and the image drawable is checked.
With this in mind if I had to refactor again I'd still probably keep all the previous `advanceUntilIdle` calls here just to move swiftly with the changes, avoid going too deep in the functionality and really avoid any regression risk.

File: components/browser/icons/src/test/java/mozilla/components/browser/icons/generator/DefaultIconGeneratorTest.kt (hunk 2 above)
Comment (Mugurell) [reply components/browser/icons/src/test/java/mozilla/components/browser/icons/generator/DefaultIconGeneratorTest.kt:49]: Thank you for catching this.
I'd say this was just a quick "replace runBlocking with runTest(UnconfinedTestDispatcher())" from my part.
Went through the code, it seems like we didn't need `runBlocking` and we don't even need `runTest` with any dispatcher here since this test is not using coroutines.
Updated as such.

File: docs/changelog.md (hunk 5 above)
Comment (Mugurell) [reply docs/changelog.md:16]: The matching Fenix changes are also ready for review and [the `runBlockingTest` usages there was indeed replaced with `runTestOnMain`](https://github.com/mozilla-mobile/fenix/pull/24892/files#diff-8c0a90dde94db153441c370ee85c29abe8402ba68915efad87a3954af5dfec6c) though they were already marked as ignored. And I kept them as such to not pollute my potential failing tests with other unrelated failures.

**Comment by Mugurell** [comment]
> when to use the plain runTest (defaulting to an empty context) vs. runTest(UnconfindedTestDispatcher()

This is indeed the biggest decision related to our unit tests.
[`runTest` does use by default `EmptyCoroutineContext`](https://github.com/Kotlin/kotlinx.coroutines/blob/194e6b659890b2ecc6c14d904e8d2d7c5bf424a9/kotlinx-coroutines-test/common/src/TestBuilders.kt#L148) and [will create a new `TestScope`](https://github.com/Kotlin/kotlinx.coroutines/blob/194e6b659890b2ecc6c14d904e8d2d7c5bf424a9/kotlinx-coroutines-test/common/src/TestBuilders.kt#L154) which [if no dispatcher is set will default to use a `StandardTestDispatcher`](https://github.com/Kotlin/kotlinx.coroutines/blob/194e6b659890b2ecc6c14d904e8d2d7c5bf424a9/kotlinx-coroutines-test/common/src/TestScope.kt#L105).

The important decision in this migration is favoring [`UnconfinedTestDispatcher`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-unconfined-test-dispatcher.html) (also set as default in the updated `MainCoroutineRule`) versus the otherwise default [`StandardTestDispatcher`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-standard-test-dispatcher.html).
The difference between these dispatchers are that the unconfined one will eagerly enter all launch and async blocks while the standard dispatcher needs calls in the scheduler to actually execute coroutines.

`UnconfinedTestDispatcher` gives no execution order guarantees. If there are more coroutines to be executed the order in which they finish is unknown and cannot be relied on. Fortunately our tests and code use small methods with few coroutines that could run concurrently. When we need to wait for one to finish before continuing the execution we can use  [`runCurrent`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-current.html) to force the tasks scheduled at this moment of time to run. So for our codebase `UnconfinedTestDispatcher` is just more straightforward to use and for this migration meant less changes and investigations.

`StandardTestDispatcher`can be more reliable because it will guarantee the order of execution (and getting the result of).
But [according to its documentation](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-standard-test-dispatcher.html):
> This [TestDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-dispatcher/index.html) instance does not itself execute any of the tasks. Instead, it always sends them to its [scheduler](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-standard-test-dispatcher.html), which can then be accessed via [TestCoroutineScheduler.runCurrent](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-coroutine-scheduler/run-current.html), [TestCoroutineScheduler.advanceUntilIdle](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-coroutine-scheduler/advance-until-idle.html), or [TestCoroutineScheduler.advanceTimeBy](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-coroutine-scheduler/advance-time-by.html), which will then execute these tasks in a blocking manner.

In most of the cases we were using `runBlocking`/`runBlockingTest` just because there was a coroutine involved but otherwise there was no need for the granularity in execution that `StandardTestDispatcher` would provide and by using `UnconfinedTestDispatcher` I could avoid adding now new `runCurrent`/`advance..` calls in all places leading to new `launch` / `async` needing to be executed.

Since I mass migrated most of the tests in which coroutines were used to rely on `UnconfinedTestDispatcher` which doesn't provide order of execution guarantees I agree there is a small risk of adding some intermittent failures but having to observe CI and fix assertions / verifications fails later is a smaller and more straightforward task.

**Comment by Mugurell** [comment]
And after all CI tasks passed following a rebase with another PR updating the changelog this needs to be rebased (and run all CI tasks) again. 😏

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt (hunk 3 above)
Comment (csadilek) [reply components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuTest.kt:437]: @Mugurell OK, I went over all the `advanceUntilIdle` calls you removed, and they do indeed all seem unnecessary! Thanks.

**Comment by csadilek** [comment]
> The difference between these dispatchers are that the unconfined one will eagerly enter all launch and async blocks while the standard dispatcher needs calls in the scheduler to actually execute coroutines.....In most of the cases we were using runBlocking/runBlockingTest just because there was a coroutine involved but otherwise there was no need for the granularity in execution that StandardTestDispatcher would provide and by using UnconfinedTestDispatcher I could avoid adding now new runCurrent/advance.. calls in all places leading to new launch / async needing to be executed.

Thank for the explanation, @Mugurell. That makes sense to me! However, I see lots of new uses of `runTest` where we don't call the scheduler to trigger execution e.g., 

- https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-d6d1891529a1c1633a54e6cff6ca0f439071c854401f05797f7fe11f297c2636R30
- https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-8b5cdfb2de0713e23c85653fa86d34ea37db608633094e84a39e9859df34875aR28
- https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-f130065a0a46dd4f5d143268d874abea2e474bb0fcd99f0f794975c42f1e8e91R42

There is still a piece I am missing? :)

File: components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt (hunk 1 above)
Comment (csadilek) [reply components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt:41]: I see how the diff would've been much easier, but also agree that `runBlockingTest` is an oversimplified name and confusing. So, OK with `runTestOnMain` to indicate that the main dispatcher is used, and therefore needs to be set in the test via `mainCorountineRule`.

File: components/browser/menu2/src/test/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHoldersTest.kt [hunk 6]
```diff
@@ -85,7 +84,7 @@ class DrawableMenuIconViewHoldersTest {
     }
 
     @Test
-    fun `async view holder sets icon on view`() = testDispatcher.runBlockingTest {
+    fun `async view holder sets icon on view`() = runTestOnMain {
```
Comment (csadilek) [inline components/browser/menu2/src/test/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHoldersTest.kt:87]: These two `runTestOnMain` here seem no needed at all, right?

**Comment by Mugurell** [comment]
I see lots of new uses of `runTest` where we don't call the scheduler to trigger execution e.g.,
> https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-d6d1891529a1c1633a54e6cff6ca0f439071c854401f05797f7fe11f297c2636R30

Here it seems it was an automatic migration on my part. 
Seems like previously `runBlocking` was used to maybe offload the preference read off of main thread. Which wouldn't really succeed in that.
I replaced it with `runTest(<StandardTestDispatcher>)` which would keep the same behavior as before. 
Still running the code on Main. Seems like a good catch to remove this and have CI run again.

&nbsp;
> https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-8b5cdfb2de0713e23c85653fa86d34ea37db608633094e84a39e9859df34875aR28

Here `runBlockingTest` was used (`runBlocking` would seem more appropriately since there aren't any delays to be skipped or other coroutines to enter eagerly) to be able to use "await" and more easily read the value from GeckoResult inside a coroutine.
`runTest` serves the same functionality - a coroutine in which to use `await` to read the value from GeckoResult.
No need to choose between the two dispatchers since there is no functionality that would by affected by them.
 
&nbsp;
> https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-f130065a0a46dd4f5d143268d874abea2e474bb0fcd99f0f794975c42f1e8e91R42
> 

Here `runBlocking` was previously used I'd say again just to have a coroutine scope in which to [call `join`](https://github.com/mozilla-mobile/android-components/blob/71e124a5dbafffb800168ea615fdbc54d7398f50/components/browser/icons/src/test/java/mozilla/components/browser/icons/extension/IconMessageHandlerTest.kt#L132) for waiting until updating the icon. [Which happens on another thread](https://github.com/mozilla-mobile/android-components/blob/71e124a5dbafffb800168ea615fdbc54d7398f50/components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessageHandler.kt#L29).

`runTest(<StandardTestDispatcher>)` works here because there are no launch/async inside it's scope that we'd need to "manually" start the execution for.

&nbsp;
-------
Had to re-read that important detail in the documentation and and went through the code again to actually remove some cases in which `UnconfinedTedDispatcher` was used incorrectly (would not cause any issue but I would expressly change the default without any benefit):
- in [ClientTest](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/concept/fetch/src/test/java/mozilla/components/concept/fetch/ClientTest.kt#L17) with the `UnconfinedTedDispatcher` we would enter the async block eagerly but there's no rush really. The below `await` will get us all we need.
- [SendTabFeatureKtTest](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/feature/accounts-push/src/test/java/mozilla/components/feature/accounts/push/SendTabFeatureKtTest.kt) tests don't contain any coroutine builders itself and doesn't pass the scope below so there's no reason to even think about what dispatcher to use.
- in [WebExtensionActionMenuCandidateTest](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/feature/addons/src/test/java/mozilla/components/feature/addons/menu/WebExtensionActionMenuCandidateTest.kt#L101-L127) a coroutine is started at line #124 but in place, without another coroutine builder being used so there's not need for `UnconfinedTedDispatcher`.
- in [AddonInstallationDialogFragmentTest] there are two tests migrated to use the `UnconfinedTedDispatcher` but again, they don't contain any coroutine builders themselves and don't pass the scope below so there's no reason to even think about what dispatcher to use.
- [RestoreUseCase - filters based on tab timeout](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/feature/tabs/src/test/java/mozilla/components/feature/tabs/TabsUseCasesTest.kt#L353-L376) would use a `UnconfinedTedDispatcher` without having a launch/async block in it's scope so there's no need to override the default.

(Will update the patch to remove these).

On the other hand some good uses of the `UnconfinedTedDispatcher` are in
- [ContainerMiddlewareTest](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/feature/containers/src/test/java/mozilla/components/feature/containers/ContainerMiddlewareTest.kt) with tests depending on a flow emission which needs the `UnconfinedTestDispatcher` for collect to actually happen with no other actions needed from us.
- [RunWhenReadyQueueTest tests](https://github.com/mozilla-mobile/android-components/blob/477f108540a5419a319d4c8afbd2603494a70f15/components/feature/addons/src/test/java/mozilla/components/feature/addons/menu/WebExtensionActionMenuCandidateTest.kt#L101-L127) which pass the scope of the `UnconfinedTedDispatcher` set in `MainCoroutineRule` to the actual implementation being tested which will `launch` new coroutines [inside RunWhenReadyQueue](https://github.com/mozilla-mobile/android-components/blob/6c0fe91ab83614265f533b819bf406fc046b2227/components/support/utils/src/main/java/mozilla/components/support/utils/RunWhenReadyQueue.kt#L38).

File: components/browser/menu2/src/test/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHoldersTest.kt (hunk 6 above)
Comment (Mugurell) [reply components/browser/menu2/src/test/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHoldersTest.kt:87]: In this tests `MainCoroutineRule` was already used.
So any `runBlocking` / `runBlockingTest` usages were directly moved to use `runTestOnMain`.

**Comment by csadilek** [comment]
Thanks for the cleanup, @Mugurell, but my question was actually about something different. In response to my question about `runTest` with the `StandardDispatcher` vs. the `UnconfinedDispatcher`, you wrote:

> The difference between these dispatchers are that the unconfined one will eagerly enter all launch and async blocks while the standard dispatcher needs calls in the scheduler to actually execute coroutines

So, I was wondering about our new use of `runTest` in the case above (with the standard dispatcher) where we don't actually haves calls to the dispatcher to execute coroutines i.e., we never call `advanceUntilIdle` and based on your description and the docs the coroutines should never run then. That's the part I am missing.

**Review by csadilek — CHANGES_REQUESTED** [review]
> So, I was wondering about our new use of runTest in the cases above (with the standard dispatcher) where we don't actually have calls to the dispatcher to execute coroutines i.e., we never call `advanceUntilIdle` and based on your description and the docs the coroutines should never run then. That's the part I am missing.

@Mugurell I read the docs again now and tried it out. So with `runTest` and the standard dispatcher the test itself will run, but nested coroutines won't launch unless `advanceUntilIdle` is called. That makes sense and is ideal for testing. 

However, even with the `UnconfinedTestDispatcher` we can't simply rely on it, as this will cause intermittent failures e.g., https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-f0cb0bdd047e2d1a322c2b71629667c0b590ba475d44a5b344cc8e908291a576R39 -> This test will actually fail intermittently. You can try it out by adding a `delay(5000)` in the `ContainerMiddleware` right before `containerStorage.addContainer`. This delay slows down the execution/completion of the coroutine (not its launch) and the test fails because we're not waiting for its completion. So, we will see intermittent failures.

We need a call to `testScheduler.advanceUntilIdle()` after https://github.com/mozilla-mobile/android-components/pull/12006/files#diff-f0cb0bdd047e2d1a322c2b71629667c0b590ba475d44a5b344cc8e908291a576R47 to fix this. So, where we have nested coroutines we need to make sure we're waiting for their completion, as they may be executed eagerly but on a different thread using `UnconfinedTestDispatcher`.

**Comment by Mugurell** [comment]
> So, I was wondering about our new use of `runTest` in the cases above (with the standard dispatcher) where we don't actually have calls to the dispatcher to execute coroutines i.e., we never call `advanceUntilIdle` and based on your description and the docs the coroutines should never run then. That's the part I am missing.

The `runCurrent` / `advance..` calls would only be needed to enter launch / async blocks started in that scope.
In the first example from the [official documentation for UnconfinedTestDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-unconfined-test-dispatcher.html) it is seen that the value of `entered` is immediately changed to `true`.
If replacing in that test `UnconfinedTestDispatcher` with `StandardTestDispatcher` then `assertTrue(entered)` would fail.
For it to pass we'd have to use something like

<pre>
<b>runCurrent() // to enter the launch block.</b>
assertTrue(entered) // because of runCurrent the value got to be updated.
assertFalse(completed) // however, the child coroutine then suspended, so it is enqueued.
deferred.complete(Unit) // resume the coroutine.
<b>advanceUntilIdle() // execute all enqueued tasks like the above `complete` call.</b>
assertTrue(completed) // now the child coroutine is immediately completed.
</pre>

**Comment by csadilek** [comment]
@Mugurell yes, please see my comment above with an example https://github.com/mozilla-mobile/android-components/pull/12006#pullrequestreview-960776443. The tests would fail intermittently with `UnconfinedTestDispatcher` so we need `advanceUntilIdle` anyway for some of them and then we can also use the `StandardTestDispatcher`.

**Comment by Mugurell** [comment]
Investigated this more with @csadilek and saw than while `StandardTestDispatcher` would have everything running on one thread it will involve a great big deal of ceremony to add now and in future tests, eg:
```
store.dispatch(
    EngineAction.LoadUrlAction(
        "test-tab",
        "https://www.firefox.com"
    )
).join()

testScheduler.advanceUntilIdle()
store.waitUntilIdle() // Wait for CreateEngineSessionMiddleware to create the new session

testScheduler.advanceUntilIdle()
store.waitUntilIdle() // Wait for the LinkingMiddleware to load url
testScheduler.advanceUntilIdle()
``` 

Using `UnconfinedTestDispatcher` doesn't eliminate this but helps reducing it consideraly resulting in almost the same code that we have now:
<pre>
store.dispatch(
   EngineAction.LoadUrlAction(
      "test-tab",
         "https://www.firefox.com"
   )
).joinBlocking()
        
dispatcher.<i>scheduler</i>.advanceUntilIdle()
store.waitUntilIdle()
</pre>

&nbsp;

The issues that were seen above show some scenarios that need more attention:
- `delay`s in SUT will allow the test to continue and finish execution. This seems like an edgecase that can be treated (by using `testScheduler.advanceTime..`) only where they are used in code. Would be applicable to both new test dispatchers.
- store actions leading to heavy calculations happening on another thread need to be explicitly waited on using `waitUntilIdle`. This is also applicable now.

File: components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    dispatchTimeoutMs: Long = DEFAULT_DISPATCH_TIMEOUT_MS,
+    testBody: suspend TestScope.() -> Unit
+): TestResult {
+    val mainDispatcher = Dispatchers.Main
+    require(mainDispatcher is TestMainDispatcher) {
+        "A TestDispatcher is not available. Use Dispatchers.setMain to set one before calling this method"
```
Comment (csadilek) [inline components/support/test/src/main/java/mozilla/components/support/test/rule/Helpers.kt:47]: I think it would be good to also point to `MainCoroutineRule` in this message.

**Review by csadilek — APPROVED** [review]
We have worked through this together and this is the best trade-off given the new API:
- `runTestOnMain` to be used in combination with `MainCoroutineRule` when nested coroutines should be entered eagerly.
- `runTest` when the dispatcher should be advanced "manually."

We also both ran this several times now and haven't seen intermittent failures. So, 🤞.

🚢 (tomorrow morning)
